packages/debug/src/browser/debug-contribution-manager.ts

```diff
@@ -0,0 +1,247 @@
+    @inject(OutputChannelManager)
+    protected readonly outputChannelManager: OutputChannelManager;
+    @inject(DebugPreferences)
+    protected readonly debugPreferences: DebugPreferences;
+
+    protected readonly onDidContributionAddEmitter = new Emitter<string>();
+    readonly onDidContributionAdd: Event<string> = this.onDidContributionAddEmitter.event;
+    protected fireDidContributionAdd(debugType: string): void {
+        this.onDidContributionAddEmitter.fire(debugType);
+    }
+
+    protected readonly onDidContributionDeleteEmitter = new Emitter<string>();
+    readonly onDidContributionDelete: Event<string> = this.onDidContributionDeleteEmitter.event;
+    protected fireDidContributionDelete(debugType: string): void {
+        this.onDidContributionDeleteEmitter.fire(debugType);
+    }
+
+    async registerDebugPluginContributor(type: string, contributor: DebugPluginContributor): Promise<Disposable> {
+        if (await this.isContributorRegistered(type)) {
+            console.warn(`Debugger with type '${type}' already registered.`);
+            return Disposable.NULL;
+        }
+
+        this.sessionContributionRegistry.registerDebugSessionContribution(type, {
+            debugType: type,
+            debugSessionFactory: () =>
+                new PluginDebugSessionFactory(
+                    this.terminalService,
+                    this.editorManager,
+                    this.breakpoints,
+                    this.labelProvider,
+                    this.messages,
+                    this.outputChannelManager,
+                    this.debugPreferences,
+                    contributor)
+        });
+
+        this.pluginContributors.set(type, contributor);
+        this.fireDidContributionAdd(type);
+        return Disposable.create(() => this.unregisterDebugPluginContributor(type));
+    }
+
+    async unregisterDebugPluginContributor(debugType: string): Promise<void> {
+        this.pluginContributors.delete(debugType);
+        this.sessionContributionRegistry.unregisterDebugSessionContribution(debugType);
+        this.fireDidContributionDelete(debugType);
+    }
+
+    async debugTypes(): Promise<string[]> {
+        const debugTypes = await this.debugService.debugTypes();
+        return debugTypes.concat(Array.from(this.pluginContributors.keys()));
+    }
+
+    async provideDebugConfigurations(debugType: string, workspaceFolderUri: string | undefined): Promise<DebugConfiguration[]> {
+        const contributor = this.pluginContributors.get(debugType);
+        if (contributor) {
+            return contributor.provideDebugConfigurations(workspaceFolderUri);
+        } else {
+            return this.debugService.provideDebugConfigurations(debugType, workspaceFolderUri);
+        }
+    }
+
+    async resolveDebugConfiguration(config: DebugConfiguration, workspaceFolderUri: string | undefined): Promise<DebugConfiguration> {
+        const contributor = this.pluginContributors.get(config.type);
```

Comment:
it is bogus, all debug providers should participate in the resolution of any `DebugConfiguration`, regardless whether they come from extensions or plugins.